syntax="proto3";

message WorkerComputeFeaturesRequest {
    string reconstruction_id = 1;
    string image_id = 2;
};

message WorkerComputeFeaturesResponse {
    bool success = 1;
};

message WorkerComputeMatchesRequest {
    string reconstruction_id = 1;
    string image_id = 2;
};

message WorkerComputeMatchesResponse{
    bool success = 1;
}

message RegisterWorkerRequest{
    string address = 1;
    int32 cores = 2;
};

message RegisterWorkerResponse {

};

message WorkerAddImageRequest {
    string reconstruction_id = 1;
    string image_id = 2;
};

message  WorkerAddImageResponse {
    bool success = 1;
};

message  WorkerIncrementalSFMRequest {
    string reconstruction_id = 1;
};

message  WorkerIncrementalSFMResponse {
    bool success = 1;
};

message  WorkerComputeStructureRequest {
    string reconstruction_id = 1;
};

message  WorkerComputeStructureResponse {
    bool success = 1;
};

service WorkerPoolManager {
    rpc Register(RegisterWorkerRequest) returns (RegisterWorkerResponse);
    rpc AddImage(WorkerAddImageRequest) returns (WorkerAddImageResponse);
    rpc ComputeFeatures(WorkerComputeFeaturesRequest) returns (WorkerComputeFeaturesResponse);
    rpc ComputeMatches(WorkerComputeMatchesRequest) returns (WorkerComputeMatchesResponse);
    rpc IncrementalSFM( WorkerIncrementalSFMRequest) returns (WorkerIncrementalSFMResponse);
    rpc ComputeStructure(WorkerComputeStructureRequest) returns (WorkerComputeStructureResponse);
}

service Worker {
    rpc AddImage(WorkerAddImageRequest) returns ( WorkerAddImageResponse);
    rpc ComputeFeatures(WorkerComputeFeaturesRequest) returns (WorkerComputeFeaturesResponse);
    rpc ComputeMatches(WorkerComputeMatchesRequest) returns (WorkerComputeMatchesResponse);
    rpc IncrementalSFM(WorkerIncrementalSFMRequest) returns (WorkerIncrementalSFMResponse);
    rpc ComputeStructure(WorkerComputeStructureRequest) returns (WorkerComputeStructureResponse);
}